Bug: sometimes getting `goblet.common_cloud_actions.MissingArtifact` in GA Actions

Hi. First of all, thanks for making and maintaining this great package! Now, with the inclusion of Cloud Run Jobs, this package makes it so easy to deploy arbitrary images for our pipeline steps, like for running `dbt` or ingesting from Cloud Storage to BigQuery. And "regular" Cloud Run for deploying APIs and Streamlit dashboards.

We use GitHub Actions for automating the `goblet deploy`, and it has worked perfectly up until I have two parallel GH Actions jobs deploying at the same time (not sure this has anything to do with the bug, but this only started occurring after having >1 `goblet deploy` on GH Actions). At first, I used the shared common default Artifact Registry called `cloud-run-source-deploy`, and I thought the error might have been related to this. So I created a dedicated Artifact Registry (AR) for each jobs Cloud Build step, but the error still continues to appear.

The error only seems to appear _sometimes_, I get a mix of successful and unsuccessful deploys via GH Actions. Sometimes it succeeds in this, sometimes not. We can't have this stochastic behavior in our production systems. I don't understand what is happening, but I suspect it could have something to do with timing, that the function [common_cloud_actions.getCloudbuildArtifact](https://github.com/goblet/goblet/blob/main/goblet/common_cloud_actions.py#L174) tries to fetch the image from Artifact Registry before Cloud Build has completed building and uploading the image. But this could also be completely wrong :D 

I am using `goblet` version 0.8.4.

Here is the GH actions step with the error produced:

```bash
Run poetry run goblet deploy --location europe-north1
INFO:goblet.app:preparing to deploy with backend CloudRun
INFO:goblet.backend:zipping source code
INFO:goblet.backend:uploading source zip to gs......
INFO:goblet.backend:source code uploaded
INFO:goblet.deployer:creating cloudbuild
INFO:goblet.backend:Skipping cloudrun deployment since it is not needed...
INFO:goblet.decorators:deploying route
INFO:goblet.decorators:deploying pubsub
INFO:goblet.decorators:deploying storage
INFO:goblet.decorators:deploying eventarc
INFO:goblet.decorators:deploying http
INFO:goblet.decorators:deploying jobs
Traceback (most recent call last):
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/common_cloud_actions.py", line 188, in getCloudbuildArtifact
    resp["results"]["images"][0]["name"]
KeyError: 'results'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/bin/goblet", line 8, in <module>
    sys.exit(main())
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/cli.py", line 79, in deploy
    app.deploy(
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/app.py", line 67, in deploy
    self.deploy_handlers(source, config=config)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/decorators.py", line 3[23](https://github.com/coopnorge/data-driven-distribution/actions/runs/3192433708/jobs/5209907201#step:9:24), in deploy_handlers
    v.deploy(source, entrypoint="goblet_entrypoint", config=config)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/resources/handler.py", line [37](https://github.com/coopnorge/data-driven-distribution/actions/runs/3192433708/jobs/5209907201#step:9:38), in deploy
    self._deploy(source, entrypoint, config=config)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/resources/jobs.py", line [51](https://github.com/coopnorge/data-driven-distribution/actions/runs/3192433708/jobs/5209907201#step:9:52), in _deploy
    artifact = getCloudbuildArtifact(self.versioned_clients.cloudbuild)
  File "/home/runner/.cache/pypoetry/virtualenvs/bucket-to-bigquery-85QkY9Aq-py3.10/lib/python3.10/site-packages/goblet/common_cloud_actions.py", line 193, in getCloudbuildArtifact
    raise MissingArtifact("Missing artifact. Cloud Build may have failed.")
goblet.common_cloud_actions.MissingArtifact: Missing artifact. Cloud Build may have failed.
Error: Process completed with exit code 1.
```

--- a/goblet/common_cloud_actions.py
+++ b/goblet/common_cloud_actions.py
@@ -171,7 +171,7 @@ class MissingArtifact(Exception):
 
 
 # calls latest build and checks for its artifact to avoid image:latest behavior with cloud run revisions
-def getCloudbuildArtifact(client, artifactName):
+def getCloudbuildArtifact(client, artifactName, config):
     defaultProject = get_default_project()
     resp = client.execute(
         "list", parent_key="projectId", parent_schema=defaultProject, params={}
@@ -179,12 +179,15 @@ def getCloudbuildArtifact(client, artifactName):
 
     # search for latest build with artifactName
     latestArtifact = None
+    build_configs = config.cloudbuild or {}
+    registry = (
+        build_configs.get("artifact_registry")
+        or f"{get_default_location()}-docker.pkg.dev/{get_default_project()}/cloud-run-source-deploy/{artifactName}"
+    )
+
     for build in resp["builds"]:
         # pending builds will not have results field.
-        if (
-            build.get("results")
-            and artifactName == build["results"]["images"][0]["name"].split("/")[-1]
-        ):
+        if build.get("results") and registry == build["results"]["images"][0]["name"]:
             latestArtifact = latestArtifact = (
                 build["results"]["images"][0]["name"]
                 + "@"

--- a/goblet/common_cloud_actions.py
+++ b/goblet/common_cloud_actions.py
@@ -171,25 +171,28 @@ class MissingArtifact(Exception):
 
 
 # calls latest build and checks for its artifact to avoid image:latest behavior with cloud run revisions
-def getCloudbuildArtifact(client):
+def getCloudbuildArtifact(client, artifactName):
     defaultProject = get_default_project()
     resp = client.execute(
         "list", parent_key="projectId", parent_schema=defaultProject, params={}
     )
-    latestBuildId = resp["builds"][0]["id"]
-    resp = client.execute(
-        "get",
-        parent_key="projectId",
-        parent_schema=defaultProject,
-        params={"id": latestBuildId},
-    )
-    try:
-        latestArtifact = (
-            resp["results"]["images"][0]["name"]
-            + "@"
-            + resp["results"]["images"][0]["digest"]
-        )
-    except KeyError:
+
+    # search for latest build with artifactName
+    latestArtifact = None
+    for build in resp["builds"]:
+        # pending builds will not have results field.
+        if (
+            build.get("results")
+            and artifactName == build["results"]["images"][0]["name"].split("/")[-1]
+        ):
+            latestArtifact = latestArtifact = (
+                build["results"]["images"][0]["name"]
+                + "@"
+                + build["results"]["images"][0]["digest"]
+            )
+            break
+
+    if not latestArtifact:
         raise MissingArtifact("Missing artifact. Cloud Build may have failed.")
     return latestArtifact
 

--- a/goblet/resources/jobs.py
+++ b/goblet/resources/jobs.py
@@ -48,7 +48,9 @@ class Jobs(Handler):
             return
 
         config = GConfig(config=config)
-        artifact = getCloudbuildArtifact(self.versioned_clients.cloudbuild, self.name)
+        artifact = getCloudbuildArtifact(
+            self.versioned_clients.cloudbuild, self.name, config=config
+        )
 
         log.info("deploying cloudrun jobs......")
         for job_name, job in self.resources.items():

--- a/goblet/resources/jobs.py
+++ b/goblet/resources/jobs.py
@@ -48,7 +48,7 @@ class Jobs(Handler):
             return
 
         config = GConfig(config=config)
-        artifact = getCloudbuildArtifact(self.versioned_clients.cloudbuild)
+        artifact = getCloudbuildArtifact(self.versioned_clients.cloudbuild, self.name)
 
         log.info("deploying cloudrun jobs......")
         for job_name, job in self.resources.items():

--- a/goblet/revision.py
+++ b/goblet/revision.py
@@ -99,7 +99,9 @@ class RevisionSpec:
         client = self.versioned_clients.run
         region = get_default_location()
         project = get_default_project_number()
-        self.latestArtifact = getCloudbuildArtifact(self.versioned_clients.cloudbuild)
+        self.latestArtifact = getCloudbuildArtifact(
+            self.versioned_clients.cloudbuild, self.name
+        )
         self.req_body = {
             "template": {
                 **self.cloudrun_revision,

--- a/goblet/revision.py
+++ b/goblet/revision.py
@@ -100,7 +100,7 @@ class RevisionSpec:
         region = get_default_location()
         project = get_default_project_number()
         self.latestArtifact = getCloudbuildArtifact(
-            self.versioned_clients.cloudbuild, self.name
+            self.versioned_clients.cloudbuild, self.name, config=self.config
         )
         self.req_body = {
             "template": {

